File: src/main/java/alfheim/client/core/handler/CardinalSystemClient.java
package alfheim.client.core.handler;

import alexsocol.asjlib.ASJUtilities;
import alexsocol.asjlib.math.Vector3;
import alfheim.api.AlfheimAPI;
import alfheim.api.spell.SpellBase;
import alfheim.client.render.world.SpellVisualizations;
import alfheim.common.core.handler.CardinalSystem.KnowledgeSystem.Knowledge;
import alfheim.common.core.handler.CardinalSystem.PartySystem.Party;
import net.minecraft.client.Minecraft;
import net.minecraft.entity.*;
import net.minecraft.entity.boss.IBossDisplayData;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.potion.Potion;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.MovingObjectPosition;
import net.minecraft.util.MovingObjectPosition.MovingObjectType;

import java.io.Serializable;
import java.util.*;

public class CardinalSystemClient {
	
	private static final Minecraft mc = Minecraft.getMinecraft();
	public static PlayerSegmentClient segment;
	
	public static PlayerSegmentClient segment() {
		if (segment == null) segment = new PlayerSegmentClient();
		return segment;
	}
	
	public static class SpellCastingSystemClient {
		
		public static void setCoolDown(SpellBase spell, int cd) {
			if (spell == null) return;
			segment().coolDown.put(spell, cd);
		}
		
		public static int getCoolDown(SpellBase spell) {
			try {
				if (spell == null) return 0;
				return segment().coolDown.get(spell);
			} catch (Throwable e) {
				System.err.println(String.format("Something went wrong getting cooldown for %s. Returning 0.", spell));
				e.printStackTrace();
				return 0;
			}
		}
		
		public static void tick() {
			try {
				for (SpellBase spell : segment().coolDown.keySet()) {
					int time = segment.coolDown.get(spell);
					if (time > 0) segment().coolDown.put(spell, time - 1);
				}
				if (segment.init > 0) --segment.init;
			} catch (Throwable e) {
				System.err.println("Something went wrong ticking spells. Skipping this tick.");
				e.printStackTrace();
			}
		}
		
		public static void reset() {
			for (SpellBase spell : segment().coolDown.keySet()) {
				segment().coolDown.put(spell, 0);
			}
		}
	}
	
	public static class TargetingSystemClient {
		
		public static boolean selectMob() {
			if (mc == null || mc.thePlayer == null) return false;
			if (segment().party == null) segment.party = new Party(mc.thePlayer);
			MovingObjectPosition mop = ASJUtilities.getMouseOver(mc.thePlayer, 128, true);
			if (mop != null && mop.typeOfHit == MovingObjectType.ENTITY && mop.entityHit instanceof EntityLivingBase) {
				if (!segment.party.isMember((EntityLivingBase) mop.entityHit)) {
					boolean invis = ((EntityLivingBase) mop.entityHit).isPotionActive(Potion.invisibility) || mop.entityHit.isInvisibleToPlayer(Minecraft.getMinecraft().thePlayer);
					if (Vector3.entityDistance(mop.entityHit, mc.thePlayer) < (mop.entityHit instanceof IBossDisplayData ? 128 : 32) && !invis) {
						segment.target = (EntityLivingBase) mop.entityHit;
						segment.isParty = false;
					}
				} else ASJUtilities.say(mc.thePlayer, "alfheimmisc.teamnotmob");
			}
			return segment.target != null && !segment.isParty;
		}
		
		public static boolean selectTeam() {
			if (mc == null || mc.thePlayer == null) return false;
			segment().isParty = true;
			if (segment.party == null) segment.party = new Party(mc.thePlayer);
			while (true) {
				EntityLivingBase team = segment.party.get(PlayerSegmentClient.partyIndex = ((++PlayerSegmentClient.partyIndex) % segment.party.count));
				if (team != null && Vector3.entityDistancePlane(mc.thePlayer, segment.target = team) < (team instanceof IBossDisplayData ? 128 : 32)) break;
			}
			return segment.isParty;
		}
	}
	
	public static class TimeStopSystemClient {
		
		public static final LinkedList<TimeStopAreaClient> tsAreas = new LinkedList<TimeStopAreaClient>();
		
		public static void stop(double x, double y, double z, Party pt, int id) {
			tsAreas.add(new TimeStopAreaClient(x, y, z, pt, id));
		}
		
		public static boolean affected(Entity e) {
			if (e == null || e instanceof IBossDisplayData) return false;
			for (TimeStopAreaClient tsa : tsAreas) {
				if (Vector3.vecEntityDistance(tsa.pos, e) < 16) {
					if (e instanceof EntityLivingBase) {
						if (!tsa.cPt.isMember((EntityLivingBase) e)) return true;
					} else {
						return true;
					}
				}
			}
			return false;
		}
		
		public static boolean affected(TileEntity te) {
			if (te == null || !te.hasWorldObj()) return false;
			for (TimeStopAreaClient tsa : tsAreas)
				if (Vector3.vecTileDistance(tsa.pos, te) < 16) return true;
			return false;
		}
		
		public static boolean inside(EntityPlayer pl) {
			for (TimeStopAreaClient tsa : tsAreas) if (Vector3.vecEntityDistance(tsa.pos, pl) < 16.5) return true;
			return false;
		}
		
		public static void render() {
			for (TimeStopAreaClient tsa : tsAreas) SpellVisualizations.redSphere(tsa.pos.x, tsa.pos.y, tsa.pos.z);
		}
		
		public static void remove(int i) {
			tsAreas.remove(new TimeStopAreaClient(0, 0, 0, null, i));
		}
		
		private static class TimeStopAreaClient {
			
			public final Vector3 pos;
			public final Party cPt;
			public final int id;
			
			public TimeStopAreaClient(double x, double y, double z, Party pt, int i) {
				pos = new Vector3(x, y, z);
				cPt = pt;
				id = i;
			}
			
			@Override
			public boolean equals(Object obj) {
				if (!(obj instanceof TimeStopAreaClient)) return false;
				return ((TimeStopAreaClient) obj).id == id;
			}
		}
	}
	
	public static class PlayerSegmentClient implements Serializable {
		
		private static final long serialVersionUID = 6871678638741684L;
		
		public final HashMap<SpellBase, Integer> coolDown = new HashMap<SpellBase, Integer>();
		public int[] hotSpells = new int[12];
		// current and max spell init time (for blue bar)
		public int init, initM;
		
		public Party party;
		public static int partyIndex;
		public EntityLivingBase target;
		public boolean isParty;
		
		public static boolean[] knowledge;
		
		public final String userName;
		
		public PlayerSegmentClient() {
			for (SpellBase spell : AlfheimAPI.spells) coolDown.put(spell, 0);
			party = new Party(mc.thePlayer);
			userName = mc.thePlayer.getCommandSenderName();
			knowledge = new boolean[Knowledge.values().length];
		}
	}
}

File: src/main/java/alfheim/client/core/handler/CardinalSystemClient.kt
package alfheim.client.core.handler

import alexsocol.asjlib.ASJUtilities
import alexsocol.asjlib.math.Vector3
import alfheim.api.spell.SpellBase
import alfheim.client.render.world.SpellVisualizations
import alfheim.common.core.handler.CardinalSystem.PartySystem.Party
import net.minecraft.client.Minecraft
import net.minecraft.entity.*
import net.minecraft.entity.boss.IBossDisplayData
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.potion.Potion
import net.minecraft.util.MovingObjectPosition.MovingObjectType
import java.io.Serializable
import java.util.*
import kotlin.collections.HashMap
import kotlin.collections.HashSet

object CardinalSystemClient {
	
	val mc get() = Minecraft.getMinecraft()!!
	
	val playerSkinsData = HashMap<String, Pair<Boolean, Boolean>>()
	
	object SpellCastingSystemClient {
		
		fun setCoolDown(spell: SpellBase?, cd: Int) {
			if (spell == null) return
			PlayerSegmentClient.coolDown[spell.name] = cd
		}
		
		fun getCoolDown(spell: SpellBase): Int {
			return try {
				PlayerSegmentClient.coolDown[spell.name] ?: 0
			} catch (e: Throwable) {
				ASJUtilities.error("Something went wrong getting cooldown for $spell. Returning 0.")
				e.printStackTrace()
				0
			}
		}
		
		fun tick() {
			try {
				for (spell in PlayerSegmentClient.coolDown.keys) {
					val time = PlayerSegmentClient.coolDown[spell]!!
					if (time > 0) PlayerSegmentClient.coolDown[spell] = time - 1
				}
				if (PlayerSegmentClient.init > 0) --PlayerSegmentClient.init
			} catch (e: Throwable) {
				System.err.println("Something went wrong ticking spells. Skipping this tick.")
				e.printStackTrace()
			}
		}
	}
	
	object TargetingSystemClient {
		
		fun selectMob(): Boolean {
			if (mc.thePlayer == null) return false
			val mop = ASJUtilities.getMouseOver(mc.thePlayer, 128.0, true)
			if (mop != null && mop.typeOfHit == MovingObjectType.ENTITY && mop.entityHit is EntityLivingBase) {
				if (PlayerSegmentClient.party?.isMember(mop.entityHit as EntityLivingBase) == false) {
					val invis = (mop.entityHit as EntityLivingBase).isPotionActive(Potion.invisibility) || mop.entityHit.isInvisibleToPlayer(Minecraft.getMinecraft().thePlayer)
					if (Vector3.entityDistance(mop.entityHit, mc.thePlayer) < (if (mop.entityHit is IBossDisplayData) 128 else 32) && !invis) {
						PlayerSegmentClient.target = mop.entityHit as EntityLivingBase
						PlayerSegmentClient.isParty = false
					}
				} else
					ASJUtilities.say(mc.thePlayer, "alfheimmisc.teamnotmob")
			}
			return PlayerSegmentClient.target != null && !PlayerSegmentClient.isParty
		}
		
		fun selectTeam(): Boolean {
			if (mc.thePlayer == null) return false
			PlayerSegmentClient.isParty = true
			
			if (PlayerSegmentClient.party == null) PlayerSegmentClient.party = Party(mc.thePlayer)
			
			while (true) {
				PlayerSegmentClient.partyIndex = ++PlayerSegmentClient.partyIndex % PlayerSegmentClient.party!!.count
				val team = PlayerSegmentClient.party!![PlayerSegmentClient.partyIndex]
				PlayerSegmentClient.target = team
				if (team != null && Vector3.entityDistancePlane(mc.thePlayer, PlayerSegmentClient.target!!) < (if (team is IBossDisplayData) 128 else 32)) break
			}
			return PlayerSegmentClient.isParty
		}
	}
	
	object TimeStopSystemClient {
		
		private val tsAreas = LinkedList<TimeStopAreaClient>()
		
		fun clear() = tsAreas.clear()
		
		fun stop(x: Double, y: Double, z: Double, pt: Party, id: Int) {
			tsAreas.add(TimeStopAreaClient(x, y, z, pt, id))
		}
		
		fun affected(e: Entity?): Boolean {
			if (e == null || e is IBossDisplayData) return false
			tsAreas
				.filter { Vector3.vecEntityDistance(it.pos, e) < 16 }
				.forEach {
					if (e is EntityLivingBase) {
						if (!it.cPt?.isMember(e as EntityLivingBase?)!!) return true
					} else {
						return true
					}
				}
			return false
		}
		
		fun inside(pl: EntityPlayer) = tsAreas.any { Vector3.vecEntityDistance(it.pos, pl) < 16.5 }
		
		fun render() {
			for (tsa in tsAreas) SpellVisualizations.redSphere(tsa.pos.x, tsa.pos.y, tsa.pos.z)
		}
		
		fun remove(i: Int) {
			tsAreas.remove(TimeStopAreaClient(0.0, 0.0, 0.0, null, i))
		}
		
		private class TimeStopAreaClient(x: Double, y: Double, z: Double, val cPt: Party?, val id: Int) {
			
			val pos = Vector3(x, y, z)
			
			override fun equals(other: Any?) = if (other !is TimeStopAreaClient) false else other.id == id
		}
	}
	
	object PlayerSegmentClient: Serializable {
		
		val coolDown = HashMap<String, Int>()
		var hotSpells = IntArray(12)
		// current and max spell init time (for blue bar)
		var init: Int = 0
		var initM: Int = 0
		
		var party: Party? = null
		var target: EntityLivingBase? = null
		var isParty: Boolean = false
		var partyIndex: Int = 0
		
		var knowledge: MutableSet<String> = HashSet()
		
		var esmAbility = true
		
		private const val serialVersionUID = 6871678638741684L
	}
}